instrumentation/anr/src/main/java/io/opentelemetry/android/instrumentation/anr/AnrDetector.java

```diff
@@ -18,45 +21,38 @@
 /** Entrypoint for installing the ANR (application not responding) detection instrumentation. */
 public final class AnrDetector {
 
-    /** Returns a new {@link AnrDetector} with the default settings. */
-    public static AnrDetector create() {
-        return builder().build();
-    }
-
-    /** Returns a new {@link AnrDetectorBuilder}. */
-    public static AnrDetectorBuilder builder() {
-        return new AnrDetectorBuilder();
-    }
-
     private final List<AttributesExtractor<StackTraceElement[], Void>> additionalExtractors;
     private final Looper mainLooper;
     private final ScheduledExecutorService scheduler;
 
-    AnrDetector(AnrDetectorBuilder builder) {
-        this.additionalExtractors = builder.additionalExtractors;
-        this.mainLooper = builder.mainLooper;
-        this.scheduler = builder.scheduler;
+    public AnrDetector(
+            List<AttributesExtractor<StackTraceElement[], Void>> additionalExtractors,
+            Looper mainLooper,
+            ScheduledExecutorService scheduler) {
+        this.additionalExtractors = additionalExtractors;
+        this.mainLooper = mainLooper;
+        this.scheduler = scheduler;
     }
 
     /**
-     * Installs the ANR detection instrumentation on the given {@link InstrumentedApplication}.
+     * Installs the ANR detection instrumentation.
      *
      * <p>When the main thread is unresponsive for 5 seconds or more, an event including the main
      * thread's stack trace will be reported to the RUM system.
      */
-    public void installOn(InstrumentedApplication instrumentedApplication) {
+    public void install(@NonNull OpenTelemetryRum openTelemetryRum) {
         Handler uiHandler = new Handler(mainLooper);
         AnrWatcher anrWatcher =
                 new AnrWatcher(
                         uiHandler,
                         mainLooper.getThread(),
-                        buildAnrInstrumenter(instrumentedApplication.getOpenTelemetrySdk()));
+                        buildAnrInstrumenter(openTelemetryRum.getOpenTelemetry()));
 
         AnrDetectorToggler listener = new AnrDetectorToggler(anrWatcher, scheduler);
         // call it manually the first time to enable the ANR detection
         listener.onApplicationForegrounded();
 
-        instrumentedApplication.registerApplicationStateListener(listener);
+        ServiceManager.get().getService(AppLifecycleService.class).registerListener(listener);
```

Comment:
I've been bitten by the service locator antipattern so so many times. I would love to discuss ways of trying to avoid doing this.

Reply:
Given that we have a means of identifying Services via the interface and determining if each should be on, is there a more explicit way of registering them in a shared way so that the registry is exposed for consumption scenarios like this, but it'll just be a simple map lookup?

Another thought: are any instrumentations and services able to access any registered service? That seems a bit chaotic, especially if certain services are optional, so we'll effectively have another dependency graph on our hands that isn't explicitly managed. Is there a way this can be closed off, or otherwise make them part of the interface the Agent exposes so they can be referenced explicitly rather than looked up?

Reply:
> effectively have another dependency graph on our hands that isn't explicitly managed

Yeah this is exactly the main issue! And it's global/static and very difficult to unweave later.